cpp/ql/lib/semmle/code/cpp/models/implementations/StdString.qll

```diff
@@ -187,26 +177,30 @@ private class StdStringAppend extends TaintFunction {
 }
 
 /**
- * The standard function `std::string.assign`.
+ * The `std::string` function `insert`.
  */
-private class StdStringAssign extends TaintFunction {
-  StdStringAssign() { this.getClassAndName("assign") instanceof StdBasicString }
+private class StdStringInsert extends StdStringTaintFunction {
+  StdStringInsert() { this.getClassAndName("insert") instanceof StdBasicString }
 
-  /**
-   * Gets the index of a parameter to this function that is a string (or
-   * character).
-   */
-  int getAStringParameterIndex() {
-    this.getParameter(result).getType() instanceof PointerType or // e.g. `std::basic_string::CharT *`
-    this.getParameter(result).getType() instanceof ReferenceType or // e.g. `std::basic_string &`
-    this.getParameter(result).getUnspecifiedType() =
-      this.getDeclaringType().getTemplateArgument(0).(Type).getUnspecifiedType() // i.e. `std::basic_string::CharT`
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // flow from string and parameter to string (qualifier) and return value
+    (
+      input.isQualifierObject() or
+      input.isParameterDeref(this.getAStringParameterIndex()) or
+      input.isParameter(this.getAnIteratorParameterIndex())
+    ) and
+    (
+      output.isQualifierObject() or
+      output.isReturnValue()
+    )
```

Comment:
If I'm understanding correctly, this is the only change that will affect behaviour - `std::string::insert` no longer has 'reverse taint flow' from the return value to the qualifier.  What is the reason for this change?  Could it be easily demonstrated in a test?

Reply:
This also replaces `output.isReturnValueDeref()` by `output.isReturnValue()`. The reason for both changes is that insert returns a fresh iterator instead of a reference to the string that is being modified (like `+=`, `append` and `replace`). So, `output.isReturnValueDeref()` is incorrect and reverse flow seems inappropriate (that's assuming I understand reverse flow correctly).

Reply:
Fixed in https://github.com/github/codeql/pull/11212/commits/bfda1f013ebbdecad3b6400d18a28708b2bd5b6f

Reply:
> What is the reason for this change?

The reason is that the "return iterator" variants are used in the dataflow taint tests, and the tests are failing on the use-use dataflow feature branch. We are missing flow because of the use of `output.isReturnValueDeref()` even when an iterator is returned.